sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotMarkEnumsWithFlags.cs

```diff
@@ -0,0 +1,130 @@
+    {
+        internal const string DiagnosticId = "S4070";
+        private const string MessageFormat = "Remove the 'FlagsAttribute' from this enum.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var enumDeclaration = (EnumDeclarationSyntax)c.Node;
+                    var enumSymbol = c.SemanticModel.GetDeclaredSymbol(enumDeclaration);
+
+                    if (!enumDeclaration.HasFlagsAttribute(c.SemanticModel) ||
+                        enumDeclaration.Identifier.IsMissing ||
+                        enumSymbol == null)
+                    {
+                        return;
+                    }
+
+                    var membersWithValues = enumSymbol.GetMembers()
+                        .OfType<IFieldSymbol>()
+                        .Select(member => new { Member = member, Value = GetEnumValueOrDefault(member) })
+                        .OrderByDescending(tuple => tuple.Value)
+                        .ToList();
+
+                    var allValues = membersWithValues.Select(x => x.Value)
+                        .OfType<ulong>()
+                        .ToList();
+
+                    var invalidMembers = membersWithValues.Where(tuple => IsInvalidFlagValue(tuple.Value, allValues))
+                        .Select(tuple => tuple.Member.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax().GetLocation())
+                        .WhereNotNull()
+                        .ToList();
+
+                    if (invalidMembers.Count > 0)
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(rule, enumDeclaration.Identifier.GetLocation(),
+                            additionalLocations: invalidMembers));
+                    }
+                }, SyntaxKind.EnumDeclaration);
+        }
+
+        private static ulong? GetEnumValueOrDefault(IFieldSymbol enumMember)
+        {
+            ulong longValue;
+            if (!enumMember.HasConstantValue ||
+                !ulong.TryParse(enumMember.ConstantValue.ToString(), out longValue))
+            {
+                return null;
+            }
+
+            return longValue;
+        }
+
+        private static bool IsInvalidFlagValue(ulong? enumValue, List<ulong> allValues)
+        {
+            return !enumValue.HasValue ||
+                (!IsZeroOrPowerOfTwo(enumValue.Value) && !IsCombinationOfOtherValues(enumValue.Value, allValues));
+        }
+
+        private static bool IsZeroOrPowerOfTwo(ulong value)
+        {
+            return (value & (value - 1)) == 0;
```

Comment:
May be worth explaining a bit or adding a link to https://stackoverflow.com/questions/600293/how-to-check-if-a-number-is-a-power-of-2